wvd-templates/DSC/Functions.ps1

```diff
@@ -327,4 +330,163 @@ function IsRDAgentRegistryValidForRegistration {
+
+    if ($sts -eq 1618) {
+        Write-Log -Message "Stopping retries for $modeAndDisplayName. The last attempt failed with Exit code=$sts"
+        if (-not $isUninstall) {
+            throw "Stopping because $modeAndDisplayName finished with Exit code=$sts"
+        }
+    }
+    else {
+        Write-Log -Message "$modeAndDisplayName finished with Exit code=$sts"
+    }
+} 
+
+<#
+.DESCRIPTION
+Uninstalls any existing RDAgent BootLoader and RD Infra Agent installations and then installs the RDAgent BootLoader and RD Infra Agent using the specified registration token.
+
+.PARAMETER AgentInstallerFolder
+Required path to MSI installer file
+
+.PARAMETER AgentBootServiceInstallerFolder
+Required path to MSI installer file
+
+.PARAMETER StartAgent
+Start the agent service (RDAgentBootLoader) immediately
+#>
+function InstallRDAgents {
+    Param(
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AgentInstallerFolder,
+        
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AgentBootServiceInstallerFolder,
+    
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [string]$RegistrationToken,
+    
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [bool]$StartAgent,
+    
+        [Parameter(mandatory = $false)]
+        [switch]$EnableVerboseMsiLogging
+    )
+
+    # Convert relative paths to absolute paths if needed
+    Write-Log -Message "Boot loader folder is $AgentBootServiceInstallerFolder"
+    $AgentBootServiceInstaller = (Get-ChildItem $AgentBootServiceInstallerFolder\ -Filter *.msi | Select-Object).FullName
+    if ((-not $AgentBootServiceInstaller) -or (-not (Test-Path $AgentBootServiceInstaller))) {
+        throw "RD Infra Agent Installer package is not found '$AgentBootServiceInstaller'"
+    }
+
+    # Convert relative paths to absolute paths if needed
+    Write-Log -Message "Agent folder is $AgentInstallerFolder"
+    $AgentInstaller = (Get-ChildItem $AgentInstallerFolder\ -Filter *.msi | Select-Object).FullName
+    if ((-not $AgentInstaller) -or (-not (Test-Path $AgentInstaller))) {
+        throw "RD Infra Agent Installer package is not found '$AgentInstaller'"
+    }
+
+    if (!$RegistrationToken) {
+        throw "No registration token specified"
+    }
+
+    RunMsiWithRetry -programDisplayName "RDAgentBootLoader" -isUninstall -argumentList @("/x {A38EE409-424D-4A0D-B5B6-5D66F20F62A5}", "/quiet", "/qn", "/norestart", "/passive") -msiOutputLogPath "C:\Users\AgentBootLoaderUnInstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging
+
+    while ($true) {
+        try {
+            $oldAgent = Get-Package -ProviderName msi -Name "Remote Desktop Services Infrastructure Agent" -ErrorAction Stop

```

Comment:
All of the existing calling code does set ErrorAction to Stop but I thought it would be good to have this so it will still work if ErrorAction  is set to some other value.

Reply:
There are other places that also rely on ErrorAction to be 'Stop' but we don't specify it to Stop
So either we set $ErrorActionPreference to Stop in the beginning and override wherever needed OR set $ErrorActionPreference to Continue and override to Stop wherever needed. I think the first would be easier as that would require the least amount changes.

If you're planning to share this code to RDAgent repo, we should set $ErrorActionPreference to Stop on top of the function so that it carries over

Reply:
I will set $ErrorActionPreference = Stop at the beginning of these new functions. I tested it and doing that will not impact the $ErrorActionPreference value in the calling code